src/os/bluestore/BlueStore.h

```diff
@@ -372,192 +463,169 @@ class BlueStore : public ObjectStore,
       } else {
 	assert(blob_bl.length());
       }
-      ::encode(blob_bl, bl);
+      bl.append(blob_bl);
     }
     void decode(bufferlist::iterator& p) {
-      ::decode(blob_bl, p);
-      undecoded = true;
+      bufferlist::iterator s = p;
+      ::decode(blob, p);
+      s.copy(p.get_off() - s.get_off(), blob_bl);
       dirty = false;
     }
   };
   typedef boost::intrusive_ptr<Blob> BlobRef;
+  typedef boost::intrusive::set<Blob> blob_map_t;
 
-  /// a map of blobs, indexed by int64_t
-  struct BlobMap {
-    typedef boost::intrusive::set<Blob> blob_map_t;
+  /// a logical extent, pointing to (some portion of) a blob
+  struct Extent : public boost::intrusive::set_base_hook<> {
+    uint32_t logical_offset = 0;      ///< logical offset
+    uint32_t blob_offset = 0;         ///< blob offset
+    uint32_t length = 0;              ///< length
+    BlobRef blob;                     ///< the blob with our data
 
-    blob_map_t blob_map;
+    explicit Extent() {}
+    explicit Extent(uint32_t lo) : logical_offset(lo) {}
+    Extent(uint32_t lo, uint32_t o, uint32_t l, BlobRef& b)
+      : logical_offset(lo), blob_offset(o), length(l), blob(b) {}
 
-    void encode(bufferlist& bl) const;
-    void decode(bufferlist::iterator& p, Cache *c);
-
-    bool empty() const {
-      return blob_map.empty();
+    // comparators for intrusive_set
+    friend bool operator<(const Extent &a, const Extent &b) {
+      return a.logical_offset < b.logical_offset;
     }
-
-    BlobRef get(int64_t id) {
-      Blob dummy(id, nullptr);
-      auto p = blob_map.find(dummy);
-      if (p != blob_map.end()) {
-	return &*p;
-      }
-      return nullptr;
+    friend bool operator>(const Extent &a, const Extent &b) {
+      return a.logical_offset > b.logical_offset;
     }
-
-    BlobRef new_blob(Cache *c) {
-      int64_t id = get_new_id();
-      Blob *b = new Blob(id, c);
-      b->get();
-      blob_map.insert(*b);
-      return b;
+    friend bool operator==(const Extent &a, const Extent &b) {
+      return a.logical_offset == b.logical_offset;
     }
 
-    void claim(BlobRef b) {
-      assert(b->id == 0);
-      b->id = get_new_id();
-      b->get();
-      blob_map.insert(*b);
+    bool blob_spans(uint32_t o, uint32_t l) {
```

Comment:
This doesn't read right by itself -- it might be the right logic in the context in which it's used. But it's not a standalone overlap (span) test, if this blob is totally to the "left" of the parameter range, it'll still return true. So either the name is misleading or something....